fs/vfs/fs_eventfd.c

```diff
@@ -0,0 +1,639 @@
+          break;
+        }
+    }
+
+  if (i >= CONFIG_EVENT_FD_NPOLLWAITERS)
+    {
+      fds->priv = NULL;
+      ret       = -EBUSY;
+      goto errout;
+    }
+
+  /* Notify the POLLOUT event if the pipe is not full, but only if
+   * there is readers.
+   */
+
+  eventset = 0;
+  if (dev->counter < (eventfd_t)-1)
+    {
+      eventset |= POLLOUT;
+    }
+
+  /* Notify the POLLIN event if the pipe is not empty */
+
+  if (dev->counter > 0)
+    {
+      eventset |= POLLIN;
+    }
+
+  if (eventset)
+    {
+      eventfd_pollnotify(dev, eventset);
+    }
+
+errout:
+  nxsem_post(&dev->exclsem);
+  return ret;
+}
+#endif
+
+int eventfd_alloc_dev(eventfd_t init_counter, int flags)
+{
+  int ret;
+  int new_fd;
+  unsigned int new_minor;
+  FAR struct eventfd_priv_s *new_dev;
+  FAR char *devpath = NULL;
+
+  /* Allocate instance data for this driver */
+
+  new_dev = eventfd_allocdev();
+  if (new_dev == NULL)
+    {
+      /* Failed to allocate new device */
+
+      return -ENOMEM;
+    }
+
+  new_dev->counter = init_counter;
+  new_dev->mode_semaphore = !!(flags & EFD_SEMAPHORE);
+
+  /* Request a unique minor device number */
+
+  new_minor = get_unique_minor();
+
+  if (new_minor < 0)
+    {
+      ferr("Cannot get minor\n");
+      ret = -EMFILE;
+      goto exit_free_new_dev;
+    }
+
+  new_dev->minor = new_minor;
+
+  /* Get device path */
+
+  if ((ret = asprintf(&devpath, EVENTFD_FMT, new_minor)) < 0)
+    {
+      ret = -ENOMEM;
+      goto exit_release_minor;
+    }
+
+  /* Register the driver */
+
+  ret = register_driver(devpath, &g_eventfd_fops, 0666, new_dev);
+  if (ret < 0)
+    {
+      ferr("Failed to register new device %s: %d\n", devpath, ret);
+      ret = -ENOMEM;
+      goto exit_free_devpath;
+    }
+
+  /* Device is ready for use */
+
+  nxsem_post(&new_dev->exclsem);
+
+  /* Try open new device */
+
+  new_fd = open(devpath, O_RDWR | (flags & (EFD_NONBLOCK | EFD_SEMAPHORE)));
+
+  if (new_fd < 0)
+    {
+      ret = -EMFILE;
+      goto exit_unregister_driver;
+    }
+
+  free(devpath);
+  return new_fd;
+
+  exit_unregister_driver:
+    unregister_driver(devpath);
+  exit_free_devpath:
+    free(devpath);
+  exit_release_minor:
+    release_minor(new_minor);
+  exit_free_new_dev:
+    eventfd_destroy(new_dev);
+  return ret;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+int eventfd(unsigned int count, int flags)
+{
+  int fd = eventfd_alloc_dev(count, flags);
+  if (fd < 0)
+    {
+      set_errno(-fd);
+      return -1;
+    }
+
+  return fd;
+}
+
+int eventfd_read(int fd, eventfd_t *value)
+{
+  return read(fd, value, sizeof (eventfd_t)) != sizeof (eventfd_t) ? -1 : 0;
+}
+
+int eventfd_write(int fd, eventfd_t value)
+{
+  return write(fd, &value,
+      sizeof (eventfd_t)) != sizeof (eventfd_t) ? -1 : 0;
+}
+
+int eventfd_get_minor(int fd)
```

Comment:
move eventfd_read/eventfd_write/eventfd_get_minor to libc only eventfd need put inside fs/vfs